### ISSUE TITLE
GraphicStroke / Mark

### ISSUE THREAD

**Issue by calinarens** [NieuwlandGeo/SLDReader#62]
I would like to implement the GraphicStroke / Graphic / Mark on LineSymbolizer.
Are you already working on this?

**Comment by ajkopinga** [comment]
That sounds like a great addition to our library!

While we're not working on this ourselves right now, we are in the process of refactoring the SLD parser. There will probably be a few merge conflicts later when you add support for GraphicStroke in the parser, but I don't think they will be difficult to fix.

Can you give us a summary of how you're planning to implement this?

**Comment by calinarens** [comment]
I will add the property for GraphicStroke and adapt the appendStyle function in OlStyler.js.
When there is a GraphicStroke element, instead of using the getLineStyle, I will split up the line "feature" into points using the SLD size and maybe dash or gap properties. Then for each point I use the existing getPointStyle to create a style and add the point to the geometry property and push it on the styles array.
The appendStyle function needs some more parameters like "resolution". I also want to see if I can get the map extent to only create styles within the current extent. I have not looked on how you cache things yet.
I will also add opacity to the marker symbols.

I did a proof of concept for the Mark part. I had some problems with ExternalGraphic because of the caching/loading mechanism.

I also have a bit of problems understanding the SLD standard regarding size and gaps between the symbols. I tried to create some sample SLD's with QGIS but sometimes it creates a <Gap> element and sometimes it uses <CssParameter name="stroke-dasharray">

**Comment by ajkopinga** [comment]
I think you can do this without modifying OlStyler.js. Maybe you can modify lineStyle.js to return a pointStyle in case of GraphicStroke and add a [GeometryFunction](https://openlayers.org/en/latest/apidoc/module-ol_style_Style.html#~GeometryFunction) to the returned point style that takes a line feature and outputs a multipoint geometry.

For the resolution, it should be possible to add this to the OlStyler function, and then in turn pass it to the getLineStyle function when it's a GraphicStroke.

An alternative approach that would avoid passing the resolution could be adding a custom [RenderFunction](https://openlayers.org/en/latest/apidoc/module-ol_style_Style.html#~RenderFunction) to the style instance. The renderfunction receives the line geometry in pixel coördinates so you don't need to know the resolution.

Clipping the marks by the map viewport will be tricky, since OL calls the stylefunction only with a feature and the resolution. I think it's best to save that optimization for another time.

**Comment by calinarens** [comment]
- I am not sure if a multipoint geometry with a single point style works, because the graphics need to be rotated with the line and therefore the style needs to be different based on each point. My plan was to calculate the direction of the point as well when converting the line.
- I tried a quick proof of concept with RenderFunction, but I have no idea how to obtain [the Vector Context](https://openlayers.org/en/latest/apidoc/module-ol_render_canvas_Immediate-CanvasImmediateRenderer.html) from the Render State.

**Comment by ajkopinga** [comment]
You can (re-)use a single point style if you set the rotation of each point on the point style before you draw it. That's how I implemented dynamic symbol rotation with a single cached point style. You can get an OL wrapper (CanvasImmediateRenderer) for the canvas in de renderer function using the ```toContext``` function from ```ol/render```.

I experimented a bit myself with a style that draws a line and marks the midpoint of each line with rotated triangles, reusing the point style and setting its rotation right before rendering.

```
const lineStyle = new Style({
  fill: new Fill(),
  stroke: new Stroke({
    color: "black",
    width: 2
  })
});

const pointStyle = new Style({
  image: new RegularShape({
    fill: new Fill({ color: "yellow" }),
    points: 3,
    radius: 8,
    stroke: new Stroke({ color: "red" })
  })
});

const lineMidpointStyle = new Style({
  renderer: (pixelCoords, renderState) => {
    const render = toContext(renderState.context);

    // Draw the line string.
    render.setFillStrokeStyle(lineStyle.getFill(), lineStyle.getStroke());
    render.drawLineString(new LineString(pixelCoords));

    // Draw the midpoint of each line segment.
    for (let k = 0; k < pixelCoords.length - 1; k++) {
      const [x1, y1] = pixelCoords[k];
      const [x2, y2] = pixelCoords[k + 1];
      const mx = 0.5 * (x1 + x2);
      const my = 0.5 * (y1 + y2);

      // Set rotation before drawing.
      const vertexStyle = pointStyle.getImage();
      vertexStyle.setRotation(Math.atan2(y2 - y1, x2 - x1) - Math.PI / 6);
      render.setImageStyle(vertexStyle);

      render.drawPoint(new Point([mx, my]));
    }
  }
});
```

**Comment by calinarens** [comment]
I have started development.
My plan was to create a doc page showing the graphicstroke style, but I am running into some problems with the docs. 
- Where is the main page that references head_custom.html and the other subpages?
- npm run docs:build fails with JSDOC_ERROR: These files do not exist: 'src/**/*.js'
- What is docker used for? It fails, because I do not have it...

**Comment by calinarens** [comment]
I have started development.

I have an issue adding "toContext" using rollup.
I don't really know rollup, but I added it like this in the config:
```
globals: {
      'ol/render': 'ol.render',
      'ol/style': 'ol.style'      
    },
```
This results though in sldreader.js starting with:
```(
function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('ol/style'), require('ol/render')) :
  typeof define === 'function' && define.amd ? define(['exports', 'ol/style', 'ol/render'], factory) :
  (global = global || self, factory(global.SLDReader = {}, global.ol.style, global.render));
}(this, (function (exports, style, render) { 'use strict';
```

So render becomes global.render instead of global.ol.render.

Any ideas?

**Comment by ajkopinga** [comment]
I had the same problem as well. Re-running ```npm run build``` fixed it for me.

**Comment by calinarens** [comment]
So, I have created a pull request. I have created a power lines page to demo the functionality. Maybe the dataset is a bit too large, because it is a bit slow. Not sure if that is OpenLayers or the renderer.

**Comment by ajkopinga** [comment]
I've reviewed your pull request.

The slowness is caused by `imageStyle.clone()` at each stroke mark render. If you leave out the `clone()` the performance improves a lot, but that breaks the rotation of the stroke marks. I think I have found a solution to that problem, but I'm going to fix it in a separate branch, to contain the scope of your pull request.

**Comment by ajkopinga** [comment]
Resolved by #64 

Thanks for your contribution! 👍

**Comment by calinarens** [comment]
Hi. When do you think you will make the next (npm) release, so I do not have the include a manually build sldreader.js?

**Comment by ajkopinga** [comment]
I think I can publish a new a release sometime next week.

**Comment by ajkopinga** [comment]
I just released v0.2.4 which contains GraphicStroke support.
